ion/binaryreader.go

```diff
@@ -36,6 +37,36 @@ func newBinaryReaderBuf(in *bufio.Reader, cat Catalog) Reader {
 	return r
 }
 
+// Reset causes the binary reader to start reading from the given input bytes
+// while skipping most of the initialization steps needed to prepare the
+// reader. Reset is most commonly called with the same bytes as the reader
+// was originally created with (e.g. via NewReaderBytes) as an optimization
+// when the same data needs to be read multiple times.
+//
+// While it is possible to call Reset with different input bytes, the Reader
+// will only work correctly if the new bytes contain the exact same binary
+// version marker and local symbols as the original input. If there are any
+// doubts whether this is the case, it is instead recommended to create a
+// new Reader using NewReaderBytes (or NewReaderCat) instead. Attempting to
+// reuse a binaryReader with inconsistent input bytes will cause the reader
+// to return errors, misappropriate values to unrelated or non-existent
+// attributes, or incorrectly parse data values.
```

Comment:
> In ion-c we have APIs to reset the underlying stream of a reader, which includes resetting the catalog and symbol table to default values.

In ion-go, you achieve this semantic by calling `ion.NewReaderBytes` or `ion.NewReader` with an io.Reader that has it's own reset capability.

> That it doesn't work "correctly" unless you provide bytes that have the exact same leading bytes also concerns me. I feel like the method would be very error prone

I tried to be clear that one has to be careful to only provide consistent bytes. Tbh, I don't expect that would be a common use case. Happy to update the comment if it needs to be more explicit.

> which leads me to think that if the application had the right information (catalog, and reader offset), it could re-create the reader itself.

Not as the library is currently written, since neither binaryReader's nor its attributes are exported, and there are no "Set" functions provided. Even if those were added, I'm not sure it would be any better than having a function like Reset that manages the details for you.